doc/dnf_daemon/examples/rpm_list_fd.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/python3
+# Copyright Contributors to the libdnf project.
+#
+# This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+#
+# Libdnf is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Libdnf is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+
+"""
+This is an example how to use `list_fd()` method of dnf5daemon-server Rpm interface.
+"""
+
+import dbus
+import json
+import os
+import select
+import time
+
+DNFDAEMON_BUS_NAME = 'org.rpm.dnf.v0'
+DNFDAEMON_OBJECT_PATH = '/' + DNFDAEMON_BUS_NAME.replace('.', '/')
+
+IFACE_SESSION_MANAGER = '{}.SessionManager'.format(DNFDAEMON_BUS_NAME)
+IFACE_RPM = '{}.rpm.Rpm'.format(DNFDAEMON_BUS_NAME)
+
+
+def repoquery(iface_rpm, options):
+    """Generator function that yields packages as they arrive from the server."""
+
+    # create a pipe and pass the write end to the server
+    pipe_r, pipe_w = os.pipe()
+    # transfer id serves as an identifier of the pipe transfer for a signal emitted
+    # after server finish. This example to not use it.
+    iface_rpm.list_fd(options, pipe_w, "the_transfer_id")
+    # close the write end - otherwise poll cannot detect the end of transmission
+    os.close(pipe_w)
+
+    # decoder that will be used to parse incomming data
+    parser = json.JSONDecoder()
+
+    # prepare for polling
+    poller = select.poll()
+    poller.register(pipe_r, select.POLLIN)
+    # wait for data 10 secs at most
+    timeout = 10000
+    # 64k is a typical size of a pipe
+    buffer_size = 65536
+
+    # remaining string to parse (can contain unfinished json from previous run)
+    to_parse = ""
+    # remaining raw data (i.e. data before UTF decoding)
+    raw_data = b""
+    while True:
+        # wait for data
+        polled_event = poller.poll(timeout)
+        if not polled_event:
+            print("Timeout reached.")
+            break
+        # we know there is only one fd registered in poller
+        descriptor, event = polled_event[0]
+        # read a chunk of data
+        buffer = os.read(descriptor, buffer_size)
+        if buffer:
```

Comment:
A suggestion - May be it would be better to switch order

```
if not buffer:
    break
else:
    ...
```

Reply:
True, `raw_data` can still contain part of input failing to `decode()`. Also `to_parse` string can contain some unfinished JSON objects. So the `repoquery` generator should probably raise an exception in case these variables are not empty after parsing loop finished.